### DISCUSSION TITLE
engine singleton Built in SQLmodel or make my own ?

### DISCUSSION THREAD

**Discussion by SteffRainville** [fastapi/sqlmodel#1510] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
Architecture question, no code needed.
```

### Description

I need the "engine" instance in my services and routers to create "session" to access the database?

   # define this somewhere
    engine = create_engine(sqlite_url, echo=False)
    
   # use this everywhere
   session = Session(engine)

All samples are single file examples so no imports of engine.

It is unclear to me if SQLModel has a solution or should I create my own singleton to hold the engine instance. 

My Flask background had me put the engine from SqlAlchemy to the app instance, is this prefred in FastApi (But not sure if I can import the app instance from main.py).

### Operating System

Windows

### SQLModel Version

0.0.4

### Python Version

3.9.7

**Comment by LucidDan** [comment, marked as answer]
If you use this style of dependency for getting the database session - https://sqlmodel.tiangolo.com/tutorial/fastapi/session-with-dependency/#create-a-fastapi-dependency

..then you only need to just create the engine in the same module as where you define the get_session() function. e.g. if you have a "db.py", you can just have something like:

```python
# ... imports, etc ...

engine = create_engine(sqlite_url, echo=False)

def get_session():
    with Session(engine) as session:
        yield session
```

Then when you define your router endpoints, just import and use get_session with Depends as described in the documentation I linked to above.

That's roughly a simplified version of how I do it, anyway...

So, not that different from the single-file examples in the tutorial, just moving the create_engine and get_session code to a separate module, and importing that.
